File: app/src/main/java/com/spbstu/sneakerhunter/adapters/FavoritesCardAdapter.java
package com.spbstu.sneakerhunter.adapters;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.spbstu.sneakerhunter.R;
import com.spbstu.sneakerhunter.server_list.Sneaker;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

public class FavoritesCardAdapter extends RecyclerView.Adapter<FavoritesCardAdapter.ViewHolder> {

    private FavoritesCardAdapter.Listener listener;
    private List<Sneaker> elements;

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private CardView cardView;

        public ViewHolder(CardView cv){
            super(cv);
            cardView = cv;
        }
    }

    private class DownloadImageTask extends AsyncTask<String, Void, Bitmap> {
        ImageView bmImage;

        public DownloadImageTask(ImageView bmImage) {
            this.bmImage = bmImage;
        }

        protected Bitmap doInBackground(String... urls) {
            String urldisplay = urls[0];
            Bitmap mIcon11 = null;
            try {
                InputStream in = new java.net.URL(urldisplay).openStream();
                mIcon11 = BitmapFactory.decodeStream(in);
            } catch (Exception e) {
                Log.e("Error", e.getMessage());
                e.printStackTrace();
            }
            return mIcon11;
        }

        protected void onPostExecute(Bitmap result) {
            bmImage.setImageBitmap(result);
        }
    }

    public interface Listener {
        void onClick(int position);
    }

    public FavoritesCardAdapter(List<Sneaker> elements) {
        this.elements = new ArrayList<>();
        this.elements = elements;
    }

    @NonNull
    @Override
    public FavoritesCardAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CardView cardView = (CardView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.favorites_cardview, parent, false);

        return new FavoritesCardAdapter.ViewHolder(cardView);
    }

    public void setListener(FavoritesCardAdapter.Listener listener) {
        this.listener = listener;
    }

    @Override
    public void onBindViewHolder(@NonNull FavoritesCardAdapter.ViewHolder holder, final int position) {
        CardView cardView = holder.cardView;
        ImageView imageView = (ImageView) cardView.findViewById(R.id.element_image);

        new FavoritesCardAdapter.DownloadImageTask(imageView).execute(elements.get(position).getPicture().getUrl());
        imageView.setContentDescription(elements.get(position).getName());

        TextView elementNameTextView = (TextView) cardView.findViewById(R.id.element_name);
        elementNameTextView.setText(elements.get(position).getName());

        TextView elementPriceTextView = (TextView) cardView.findViewById(R.id.element_price);
        elementPriceTextView.setText(elements.get(position).getMoney());

        ImageButton elementDeletionButton = (ImageButton) cardView.findViewById(R.id.element_deletion);


        cardView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.onClick(position);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return elements.size();
    }
}

File: app/src/main/java/com/spbstu/sneakerhunterkotlin/adapters/FavoritesCardAdapter.kt
package com.spbstu.sneakerhunterkotlin.adapters

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.AsyncTask
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.spbstu.sneakerhunterkotlin.R
import com.spbstu.sneakerhunterkotlin.server_list.Sneaker
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.BufferedInputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.*


class FavoritesCardAdapter(elementsParam: Map<Int, Sneaker>) :
    RecyclerView.Adapter<FavoritesCardAdapter.ViewHolder>() {
    private var listener: Listener? = null
    private var deleteElementListener: Listener? = null
    private var elements: Map<Int, Sneaker> = elementsParam

    class ViewHolder(val cardView: CardView) : RecyclerView.ViewHolder(cardView)

    private fun  loadImage(urlStr: String?): Bitmap? {
        val urldisplay = "https://$urlStr"
        val url = URL(urldisplay)
        val urlConnection = url.openConnection() as HttpURLConnection
        var bitmap: Bitmap? = null
        try {
            val inputStream = BufferedInputStream(urlConnection.inputStream)
            bitmap = BitmapFactory.decodeStream(inputStream)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            urlConnection.disconnect()
        }
        return bitmap
    }

//    fun setNewList(list: List<Sneaker>) {
//        elements = list.toMutableList()
//        notifyDataSetChanged()
//    }

    private fun executeLoadingImage(urlStr: String?, imageView: ImageView) {
        if (urlStr != null) {
            var result: Bitmap? = null
            GlobalScope.launch {
                withContext(Dispatchers.IO){
                    result = loadImage(urlStr)
                }
                withContext(Dispatchers.Main){
                    imageView.setImageBitmap(result)
                }
            }
        }
    }

    interface Listener {
        fun onClick(position: Int)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val cardView = LayoutInflater.from(parent.context)
            .inflate(R.layout.favorites_cardview, parent, false) as CardView
        return ViewHolder(cardView)
    }

    fun setListener(listener: Listener?, deleteElementListener: Listener?) {
        this.listener = listener
        this.deleteElementListener = deleteElementListener
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val cardView = holder.cardView

        val imageView = cardView.findViewById<View>(R.id.element_image) as ImageView
        executeLoadingImage(elements[position]?.picture?.url, imageView)
        imageView.contentDescription = elements[position]?.name

        val elementNameTextView = cardView.findViewById<View>(R.id.element_name) as TextView
        elementNameTextView.text = elements[position]?.name

        val elementPriceTextView = cardView.findViewById<View>(R.id.element_price) as TextView
        elementPriceTextView.text = elements[position]?.money

        val elementDeletionButton = cardView.findViewById<View>(R.id.element_deletion) as ImageButton
        elementDeletionButton.setOnClickListener {
            deleteElementListener?.onClick(position)
        }

        cardView.setOnClickListener {
            listener?.onClick(position)
        }
    }

    override fun getItemCount(): Int {
        return elements.size
    }

}